src/linear_algebra/dual.lean

```diff
@@ -398,3 +407,192 @@ begin
+
+variables {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M] [module R M]
+variable {W : submodule R M}
+
+/-- The `dual_annihilator` of a submodule `W` is the set of linear maps `φ` such
+  that `φ w = 0` for all `w ∈ W`. -/
+def dual_annihilator {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M]
+  [module R M] (W : submodule R M) : submodule R $ module.dual R M :=
+{ carrier := { φ | ∀ w ∈ W, φ w = 0 },
+  zero_mem' := by simp,
+  add_mem' :=  by { intros φ ψ hφ hψ w hw,
+    rw [linear_map.add_apply, hφ w hw, hψ w hw, add_zero] },
+  smul_mem' := by { intros c φ hφ w hw,
+    rw [linear_map.smul_apply, hφ w hw, smul_zero] } }
+
+@[simp] lemma mem_dual_annihilator (φ : module.dual R M) :
+  φ ∈ W.dual_annihilator ↔ ∀ w ∈ W, φ w = 0 := iff.rfl
+
+/-- The `dual_restrict` of a submodule `W` of `M` is the linear map from the
+  dual of `M` to the dual of `W` such that the domain of each linear map is
+  restricted to `W`. -/
+def dual_restrict (W : submodule R M) :
+  module.dual R M →ₗ[R] module.dual R W :=
+linear_map.dom_restrict' W
+
+@[simp] lemma dual_restrict_apply
+  (W : submodule R M) (φ : module.dual R M) (x : W) :
+  W.dual_restrict φ x = φ x.1 := rfl
+
+lemma dual_restrict_ker_eq_dual_anihilator
+  (W : submodule R M) : W.dual_restrict.ker = W.dual_annihilator :=
+begin
+  ext φ, split; intro hφ,
+  { intros w hw,
+    rw linear_map.mem_ker at hφ,
+    rw [← W.dual_restrict_apply φ ⟨w, hw⟩, hφ], refl },
+  { ext, exact hφ x.1 x.2 }
+end
+
+end submodule
+
+namespace subspace
+
+open submodule linear_map
+
+universes u v w
+
+-- We work in vector spaces because `exists_is_compl` only hold for vector spaces
+variables {K : Type u} {V : Type v} [field K] [add_comm_group V] [vector_space K V]
+
+/-- Given a subspace `W` of `V` and an element of its dual `φ`, `dual_lift W φ` is
+  the natural extenstion of `φ` to an element of the dual of `V`. -/
+noncomputable def dual_lift
+  (W : subspace K V) (φ : module.dual K W) : module.dual K V :=
+let h := classical.indefinite_description _ W.exists_is_compl in of_is_compl h.2 φ 0
+
+variable {W : subspace K V}
+
+@[simp] lemma dual_lift_of_subtype {φ : module.dual K W} (w : W) :
+  W.dual_lift φ (w : V) = φ w :=
+by erw of_is_compl_left_apply _ w
+
+lemma dual_lift_of_mem {φ : module.dual K W} {w : V} (hw : w ∈ W) :
+  W.dual_lift φ w = φ ⟨w, hw⟩ :=
+dual_lift_of_subtype ⟨w, hw⟩
+
+@[simp] lemma dual_lift_zero : W.dual_lift 0 = 0 := by simp [dual_lift]
+
+@[simp] lemma dual_lift_add (φ ψ : module.dual K W) :
+  W.dual_lift (φ + ψ) = W.dual_lift φ + W.dual_lift ψ :=
+begin
+  -- `change` is significantly slower than `show`
+  show of_is_compl _ _ 0 = of_is_compl _ φ 0 + of_is_compl _ ψ 0,
+  rw [← zero_add (0 : _ →ₗ[K] _), of_is_compl_add], simp
+end
+
+@[simp] lemma dual_lift_smul (c : K) (φ : module.dual K W) :
+  W.dual_lift (c • φ) = c • W.dual_lift φ :=
+begin
+  show of_is_compl _ _ 0 = c • of_is_compl _ _ 0,
+  rw [← smul_zero c, of_is_compl_smul], simp
+end
```

Comment:
These three lemmas could be merged into a single `dual_lift` defined as `→ₗ[R]`.

Reply:
I've added `dual_lift'` which is the linear map version of `dual_lift`

Reply:
I think it would be better to have only the linear map version, because that way it's not awkard to apply lemmas like `dual_lift'.map_sum` when the goal contains `dual_lift` instead of `dual_lift'`.

Reply:
Plus, you don't need `dual_lift_zero` because the linear_map API proves that for you.

Reply:
I've removed the non-linear map version now